src/gpl2/include/gpl2/DgReplace.h

```diff
@@ -0,0 +1,212 @@
+///////////////////////////////////////////////////////////////////////////
+//
+// BSD 3-Clause License
+//
+// Copyright (c) 2023, Google LLC
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+namespace odb {
+class dbDatabase;
+class dbInst;
+}  // namespace odb
+namespace sta {
+class dbSta;
+class dbNetwork;
+}  // namespace sta
+
+namespace grt {
+class GlobalRouter;
+}
+
+namespace rsz {
+class Resizer;
+}
+
+namespace utl {
+class Logger;
+}
+
+namespace gpl2 {
+
+class PlacerBaseCommon;
+class PlacerBase;
+class RouteBase;
+class TimingBase;
+class InitialPlace;
+class NesterovPlace;
+
+class DgReplace
+{
+ public:
+  DgReplace();
+  ~DgReplace();
+
+  void init(sta::dbNetwork* network,
+            odb::dbDatabase* odb,
+            rsz::Resizer* resizer,
+            grt::GlobalRouter* router,
+            utl::Logger* logger);
+  void reset();
+
+  // The three main functions
+  void doInitialPlace();
+  int doNesterovPlace(int start_iter = 0);
```

Comment:
I only see two

Reply:
Fixed the comment, it was here since the CUDA-native implementation, there's no clear indication what would've been the third function.